src/credentials/FabricTable.cpp

```diff
@@ -793,12 +814,89 @@ CHIP_ERROR FabricTable::Init(PersistentStorageDelegate * storage)
     // iterator doesn't have mechanism to load fabric info from storage on demand.
     // TODO - Update ConstFabricIterator to load fabric info from storage
     static_assert(kMaxValidFabricIndex <= UINT8_MAX, "Cannot create more fabrics than UINT8_MAX");
-    for (FabricIndex i = kMinValidFabricIndex; i <= kMaxValidFabricIndex; i++)
+
+    mFabricCount = 0;
+    for (auto & fabric : mStates)
     {
-        FabricInfo * fabric = &mStates[i - kMinValidFabricIndex];
-        if (LoadFromStorage(fabric) == CHIP_NO_ERROR)
+        fabric.Reset();
+    }
+    mNextAvailableFabricIndex.SetValue(kMinValidFabricIndex);
+
+    uint8_t buf[IndexInfoTLVMaxSize()];
+    uint16_t size = sizeof(buf);
+    DefaultStorageKeyAllocator keyAlloc;
+    CHIP_ERROR err = mStorage->SyncGetKeyValue(keyAlloc.FabricIndexInfo(), buf, size);
+    if (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
+    {
+        // No fabrics yet.  Nothing to be done here.
+    }
+    else
+    {
+        ReturnErrorOnFailure(err);
+        TLV::ContiguousBufferTLVReader reader;
+        reader.Init(buf, size);
+
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
+        TLV::TLVType containerType;
+        ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
```

Comment:
This entire else block is huge. Could it be factored  in a private/protected method that returns CHIP_ERROR?

Reply:
Yes, absolutely.  Will do.